### ISSUE TITLE
Vector field space of complex space

### ISSUE THREAD

**Issue by chongchenmath** [odlgroup/odl#693]
I met with an interesting phenomenon:

``space = odl.uniform_discr(min_pt=[-16, -16], max_pt=[16, 16], shape=[256, 256], dtype='float32', interp='linear')``
``shifts = [not s % 2 for s in space.shape]``
``ft_op = odl.trafos.FourierTransform(space, halfcomplex=False, shift=shifts)``

The `dtype` is different below:

``ft_op.range.vector_field_space``
Output: ``ProductSpace(uniform_discr([-25.230915999143026, -25.230915999143026], [25.034566458293664, 25.034566458293664], [256, 256], dtype='float32'), 2)``

``odl.ProductSpace(ft_op.range, ft_op.range.ndim)``
Output: ``ProductSpace(uniform_discr([-25.230915999143026, -25.230915999143026], [25.034566458293664, 25.034566458293664], [256, 256], dtype='complex64'), 2)``

Why?

**Comment by adler-j** [comment]
`ft_op.range.vector_field_space` is supposed to be the tangent bundle (ish) of the space `ft_op.range`, while `odl.ProductSpace(ft_op.range, ft_op.range.ndim)` would be the set of (`ft_op.range.ndim`) vectors on the range. This was a concious design decision at some point, see the discussion about it [here](https://github.com/odlgroup/odl/pull/488#discussion-diff-70392722).

**Comment by chongchenmath** [comment]
The following is the definition of `vector_field_space`.

````
@property
    def vector_field_space(self):
        """Space of vector fields based on ``self``."""
        return ProductSpace(self.real_space, self.ndim)
````
It just supports `real_space`, not `complex_space`. I think that is the reason. Right?

**Comment by adler-j** [comment]
It explicitly wants the real space, that was one of the motivating reasons behind introducing it.

Otherwise tests like [this test](https://github.com/odlgroup/odl/blob/master/odl/test/deform/linearized_deform_test.py#L35-L37) would fail with complex dtypes.

**Comment by adler-j** [comment]
I guess the naming is the big issue at hand here. Perhaps it should be renamed to `tangent_boundle`. Thoughts @kohr-h, @ozanoktem?

**Comment by ozanoktem** [comment]
I'm trying to follow the discussion. First, what is @chongchenmath trying to do mathematically? Is it to define a vector space of vector fields over the complex number field? In such case we are dealing with the tangent bundle and an appropriate terminology would be `tangent_bundle` as proposed by @adler-j . 

Mathematics background: ODL can only handle manifolds M that have a trivial tangent bundle, i.e. M has a tangent bundle that can be expressed as a Cartesian product. This is e.g. the case for M=R^n or M=S^1, whereas the unit sphere S^2 has a non-trivial tangent bundle (can be proud using the hairy ball theorem).

**Comment by chongchenmath** [comment]
@ozanoktem I certainly  wanted to define a space of vector fields over the complex number field, such as the range of Fourier transform. Geometric speaking, a vector field on manifold M is defined as a continuous map `X : M --> TM`, i.e., `p --> X_p`, where `X_p \in T_pM` for each `p \in M`, and `TM` is the tangent_bundle. If `M = R^n`, then `T_pM = R^n` and `TM = R^n \times R^n`. So here the space of vector fields we talked can be seen as the tangent_bundle. @adler-j @kohr-h

**Comment by chongchenmath** [comment]
A simple analysis:
````
space = odl.uniform_discr(min_pt=[-16, -16], max_pt=[16, 16], shape=[256, 256], 
dtype='float32', interp='linear')
````
is the space of `{f | f : \domain --> R^1}`.

The `vector_field_space` or `tangent bundle` of `\domain` is `{g | g : \domain --> R^2} = odl.ProductSpace(space, space.ndim)`. This is reasonable.

Note that `vector_field_space` belongs to `\domain`. So `space.vector_field_space` doesn't make sense. It should be replaced by `space.domain.vector_field_space`? @adler-j  @kohr-h @ozanoktem

**Comment by adler-j** [comment]
> @ozanoktem I certainly  wanted to define a space of vector fields over the complex number field, such as the range of Fourier transform. Geometric speaking, a vector field on manifold M is defined as a continuous map `X : M --> TM`, i.e., `p --> X_p`, where `X_p \in T_pM` for each `p \in M`, and `TM` is the tangent_bundle. If `M = R^n`, then `T_pM = R^n` and `TM = R^n \times R^n`. So here the space of vector fields we talked can be seen as the tangent_bundle.

It seems you are correct, the tangent bundle of a complex space is itself complex.

> A simple analysis:

> space = odl.uniform_discr(min_pt=[-16, -16], max_pt=[16, 16], shape=[256, 256], 
dtype='float32', interp='linear')
is the space of {f | f : \domain --> R^1}.

> The vector_field_space or tangent bundle of \domain is {g | g : \domain --> R^2} = odl.ProductSpace(space, space.ndim). This is reasonable.


The real setting is indeed trivial since the tangent bundle and the space of vectors coincide.

> Note that vector_field_space belongs to \domain. So space.vector_field_space doesn't make sense. It should be replaced by space.domain.vector_field_space?

There are two concepts at play here. The continuous `vector_field_space`, which we do not represent anywhere, and its discretization. `space.vector_field_space` is the discretization, and hence needs to be a property of the `space` since `space.domain` is not discretized and contains no information on discretization points etc.

### My conclusion

With that said, it seems to me that we need to do a renaming here. First, I'd suggest creating a property `tangent_bundle` which does what @chongchenmath is suggesting here:

```
@property
def tangent_bundle(self):
    return ProductSpace(self, self.ndim)
```

Then we need to change all current instances of `space.vector_field_space`, where it is used for discretizations, to `space.real_space.tangent_bundle`. In this way, everything should be consistent and nicely defined mathematically, right?

**Comment by kohr-h** [comment]
While this is all correct (I guess), my feeling is that a name like `tangent_bundle` is of very little help for anybody not familiar with differential geometry. And unless our ambition is to show off mathematically, we should pick a name that people have a chance to understand.
The more down-to-earth version of what we want is "the real version of the range of the gradient defined on X", mainly to define displacement fields x --> x + v(x) (or "location fields" x --> u(x)). Could you describe what exactly this space is needed for apart from that, @adler-j, @chongchenmath?

**Comment by kohr-h** [comment]
Oh, I was late. Anyway, we seem to need two things: the tangent bundle of `X` and the tangent bundle of the real version of `X`. But actually, we only happen to need the last thing because it happens to coincide with the "real tangent bundle". If we one day introduce less trivial manifolds, we will still need the "location field" space, while there won't be a use for the "real tangent bundle". I don't know the case for the proper tangent bundle, though.

**Comment by adler-j** [comment]
> Oh, I was late. Anyway, we seem to need two things: the tangent bundle of X and the tangent bundle of the real version of X. But actually, we only happen to need the last thing because it happens to coincide with the space of "location fields". If we one day introduce less trivial manifolds, we will still need the "location field" space, while there won't be a use for the "real tangent bundle". 

It is more complicated than that. Even if we did introduce less trivial manifolds, the "real tangent bundle" would still be useful since it provides a way to generate the "location field" via the [exponential map](https://en.wikipedia.org/wiki/Exponential_map_(Riemannian_geometry)), this is the formalism used to construct linear deformations etc.

> I don't know the case for the proper tangent bundle, though.

This is basically used to define a range for the gradient, I don't know any other use, but perhaps @chongchenmath has one?

**Comment by chongchenmath** [comment]
I am a little bit confused. 

Let me first analyze the interesting phenomenon at the beginning.  

The actual meaning of `ft_op.range.vector_field_space` in ODL is the vector field space of `ft_op.range.domain`, i.e., `ft_op.range.domain.vector_field_space` (just for understanding, even not defined in ODL). The `ft_op.range.domain` is `IntervalProd([-16.0, -16.0], [16.0, 16.0])`, i.e., a `\domain` on `R^2`. So the output of `ft_op.range.vector_field_space` coincides with `ProductSpace(uniform_discr([-25.230915999143026, -25.230915999143026], [25.034566458293664, 25.034566458293664], [256, 256], dtype='float32'), 2)`. The `dtype` is right. 

In contrast, `odl.ProductSpace(ft_op.range, 2)` means a product space `ft_op.range \times ft_op.range`. So its output is `ProductSpace(uniform_discr([-25.230915999143026, -25.230915999143026], [25.034566458293664, 25.034566458293664], [256, 256], dtype='complex64'), 2)`. The `dtype` is also right.

In conclusion, perhaps we don't need to change `vector_field_space`.

**Comment by adler-j** [comment]
> In conclusion, perhaps we don't need to change vector_field_space.

Perhaps, but I feel that the naming is sub-optimal since it obviously causes confusion, even when used by competent users. On the other hand, Holger is correct in that `tangent_bundle` is "heavy", but then again, do all methods really need to be simple to understand? I guess most users would not be using this method at all.

**Comment by chongchenmath** [comment]
> Perhaps, but I feel that the naming is sub-optimal since it obviously causes confusion, even when used by competent users. On the other hand, Holger is correct in that tangent_bundle is "heavy", 

Agree. We can choose `tangent_bundle` or `vector_field_space`.

**Comment by kohr-h** [comment]
Yeah, well, I don't have a strong opinion here, so go with whatever seems most appropriate to you guys.
